torch/csrc/autograd/engine.cpp

```diff
@@ -480,16 +484,21 @@ static variable_list call_function(NodeTask& task) {
 auto Engine::evaluate_function(NodeTask& task) -> void {
   // If exec_info_ is not empty, we have to instrument the execution
   auto & exec_info_ = task.base_->exec_info_;
-  if (!exec_info_.empty()) {
-    auto & fn_info = exec_info_.at(task.fn_.get());
-    if (auto *capture_vec = fn_info.captures_.get()) {
-      // Lock mutex for writing to task.base_->captured_vars_
-      std::lock_guard<std::mutex> lock(task.base_->mutex_);
-      for (auto capture : *capture_vec) {
-        task.base_->captured_vars_[capture.output_idx_] = task.inputs_[capture.input_idx_];
+  {
+    // Lock mutex for writing to task.base_->captured_vars_ and looking up
+    // exec_info_.
+    std::lock_guard<std::mutex> lock(task.base_->mutex_);
```

Comment:
Hmm, what's going on here? I see the updated comment, but I don't know why it used to be OK to access `exec_info_` without a lock and now it's not. (Or is this a bugfix?)

Reply:
> I'm assuming the code had the invariant of read-only exec_info_ for some reason.

Not sure what you mean by this :)

Reply:
Hmm, if we follow my suggestion at https://github.com/pytorch/pytorch/pull/27940#discussion_r334992862 you don't have to write to `exec_info_` anymore. So then we can undo this!